Added win_http_proxy and win_inet_proxy

jborean93 opened the pull request:
##### SUMMARY
Adds 2 modules that can be used to configure proxy settings on Windows;

* win_http_proxy - Configures proxy settings for WinHTTP
* win_inet_proxy - Configures proxy settings for WinINet (IE)

While the input options are similar, WinINet offers a few extra options and WinHTTP can also import from WinINet. This made it difficult to have just 1 `win_proxy` module and so I created a module for each for a cleaner UI. Hopefully the documentation makes this clear as to that they do and link to each other.

Fixes https://github.com/ansible/ansible/issues/22663

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_http_proxy
win_inet_proxy

jborean93 committed:
Added win_http_proxy and win_inet_proxy

jborean93 committed:
Fix up docs sanity issues

jborean93 committed:
removed duplicate doc entry

johnboy2 commented on lib/ansible/modules/windows/win_http_proxy.ps1:
```diff
@@ -0,0 +1,264 @@
+    bypass = $actual_proxy.ProxyBypass
+}
+
+if ($source -eq "ie") {
```
IIRC, the underlying "import from IE" logic provided by Microsoft silently ignores the autoconfig URL -- which seems to be a mistake to me. Do we want to propagate that behavior? I suspect a warning would be better.

johnboy2 replied:
(Admittedly, I haven't actually looked at that in a very long time, so I could be mis-remembering.)

jborean93 replied:
I couldn't find any way to actually set an auto config url in netsh/winhttp so I think it is a limitation of the latter? Could be wrong but the [WINHTTP_PROXY_INFO](https://docs.microsoft.com/en-us/windows/desktop/api/Winhttp/ns-winhttp-__unnamed_struct_3) structure does not have a field for this.

This sounds like more of a documentation thing with the potential of adding this in as a separate param if it is possible.

johnboy2 replied:
Sorry, I guess I wasn't clear. The "mistake" I was referring to was the fact that netsh doing this type of import **_silently_** ignores the auto-config URL (IIRC).

I'm not surprised that there's no winhttp option for this, since that would imply winhttp.dll having a JS interpreter (which doesn't match well with its use-case). From that perspective, _ignoring_ the auto-config URL when doing an import is entirely reasonable on MS' part -- **_except_** if it is done silently. IMHO that should really be a warning.

jborean93 replied:
Hmm I might be missing something, it's early morning here. Just to double check you are saying that the netsh import doesn't use the auto config value at all? That's the assumption that I was under. The trouble with adding warnings to modules is that it will automatically display when being used. This case is more suited towards documentation where people reading the options should be aware that `source: ie` will just import the proxy URL and bypass list.

johnboy2 replied:
RE not used at all -- correct. (I'm still going from memory on that one, but that is how I remember it.) More specifically, if you import from IE, which _only_ is set to use a [valid] auto-config script, you'll get nothing whatsoever in the winhttp settings.

As far warning or doc, my _preference_ would be to do both, but I won't get up on a high horse about it either way. :-)

johnboy2 replied:
(Should probably actually double-check my recollection on this first though. We'd all feel really silly if I'm remembering wrong.)

jborean93 replied:
Cool so yes the import is only for explicit proxy URLs and bypass list. I'll update the docs to further clarify this but I don't think a warning is necessary. This is up to the user making sure they understand the behaviour and read the docs.

johnboy2 replied:
Good enough for me.

jborean93 replied:
I also found [WinHttpGetProxyForUrl](https://docs.microsoft.com/en-us/windows/desktop/api/winhttp/nf-winhttp-winhttpgetproxyforurl) which seems to be the mechanism for WinHTTP applications to support auto config urls. This indicates it's up to the application to look this up and set on their requests and not a system wide setting.

https://docs.microsoft.com/en-us/windows/desktop/WinHttp/winhttp-autoproxy-api for some more info.

johnboy2 replied:
Yeah, your second link is pretty succinct on the matter:
> AutoProxy support is not fully integrated into the HTTP stack in WinHTTP. Before sending a request, the application must call ...

In that respect, your doc content changes look good.

johnboy2 commented on lib/ansible/modules/windows/win_http_proxy.py:
```diff
@@ -0,0 +1,103 @@
+version_added: '2.8'
+short_description: Manages proxy settings for WinHTTP
+description:
+- Used to set, remote, or import proxy settings for Windows HTTP Services
```
`remote` -> `remove`?

jborean93 replied:
Thanks, fixed.

johnboy2 commented on lib/ansible/modules/windows/win_http_proxy.py:
```diff
@@ -0,0 +1,103 @@
+    - This is like running C(netsh winhttp import proxy source=ie).
+    - The value is imported when the module runs and will not automatically
+      be updated if IE changes the values in the future. The module will have
+      to be run again to sync the latest changes.
```
Nitpick - the term `The value` is used twice here with different meanings each time.

I suggest changing `...if IE changes the values...` to something like `...if the IE configuration changes...` to resolve the ambiguity.

jborean93 replied:
Thanks sounds a lot better.

johnboy2 commented on lib/ansible/modules/windows/win_http_proxy.py:
```diff
@@ -0,0 +1,103 @@
+    type: str
+notes:
+- This is not the same as the proxy settings set in Internet Explorer, also
+  known as C(WinINet), use the M(win_inet_proxy) module to manage that instead.
```
Grammar nitpick... Quick fix here would be to swap the comma for a semi-colon.

The more detailed version is that this forms an invalid dependent clause the way it is written right now. A semi-colon would fix that, as would splitting off the `use the ...` content into its own sentence.

jborean93 replied:
I knew there was a reason I preferred Maths over English in school :)

johnboy2 commented on lib/ansible/modules/windows/win_http_proxy.py:
```diff
@@ -0,0 +1,103 @@
+    description:
+    - A string or dict that specifies the proxy to be set.
+    - If setting a string, should be in the form C(hostname), C(hostname:port),
+      or C(protocol=hostname:port).
```
Should we document that default per-protocol ports get used if one isn't set explicitly? (This is one of those things that should be obvious, except to everyone for whom it isn't.)

jborean93 replied:
Doesn't hurt, bound to get asked or someone would be interested to know.

johnboy2 commented on lib/ansible/modules/windows/win_http_proxy.ps1:
```diff
@@ -0,0 +1,264 @@
+
+if ($bypass) {
+    $bypass = $bypass -join ';'
+}
```
Should probably fail if `$bypass` is set to something but `$proxy` is not. (The `required_by` check isn't that strict.)

jborean93 replied:
Not sure what you mean, `required_by` does exactly that and it's part of the integration tests.

johnboy2 replied:
I can double-check, but I thought setting proxy to an empty string would break the check.

jborean93 replied:
Ah yes that's a good point, forgot that `proxy: ''` means no proxy at all. Will fix up.

johnboy2 commented on lib/ansible/modules/windows/win_inet_proxy.py:
```diff
@@ -0,0 +1,154 @@
+short_description: Manages proxy settings for WinINet and Internet Explorer
+description:
+- Used to set, remote, or import proxy settings for Windows INet which includes
+  Internet Explorer.
```
`Used to set or remove proxy settings...`
* `remote` -> `remove`
* There is no importing IE settings here

johnboy2 commented on lib/ansible/modules/windows/win_inet_proxy.py:
```diff
@@ -0,0 +1,154 @@
+- name: Import IE proxy configuration to WinHTTP
+  win_http_proxy:
+    source: ie
+'''
```
Suggest adding a combined example that also uses win_credential.

jborean93 replied:
Good idea, I forgot to actually test it out so I'm doing that right now.

johnboy2 reviewed (commented):
Looks great overall!

I should be able to do some testing in my environment over the next week or so, so this is purely a code review for now.

jborean93 committed:
Fix docs issues and fix for empty proxy

jborean93 commented:
Thanks a lot for looking through this @johnboy2. I've made the changes requested and have also come across a bug in win_credential which I will fix outside of this PR. Hopefully I've covered everything you've brought up.

jborean93 commented:
Fix for win_credential https://github.com/ansible/ansible/pull/54695.

johnboy2 commented on lib/ansible/modules/windows/win_http_proxy.py:
```diff
@@ -0,0 +1,103 @@
+      just the hostname.
+    - Use C(<-loopback>) to stop automatically bypassing the proxy when
+      connecting through any loopback address like C(127.0.0.1), C(localhost),
+      or the local hostname.
```
Did some testing, but was _unable to validate_ this `<-loopback>` part, at least for winhttp.

In particular, on W2016, localhost and 127.0.0.1 destinations appear to ignore the proxy unconditionally, whether `<-loopback>` is in the bypass list or not. I'm also not sure my use of W2016 is a factor, since the MS docs seem to indicate it hasn't materially changed since Vista/2008. I have also not found any support for the idea that `<-loopback>` does anything in the MS docs I reviewed.

Note that I did [find a reference](https://docs.microsoft.com/en-us/windows/desktop/api/wininet/nf-wininet-internetopena) for this construct for wininet, but not for winhttp.

Is this a real thing?

jborean93 replied:
It's definitely there for WinINet and I just assumed WinHTTP was the same but it seems like this isn't the case.

As you said I found the reference to `<-loopback>` in the `InternetOpen` function as well as https://blogs.msdn.microsoft.com/ieinternals/2013/10/11/understanding-web-proxy-configuration/. This then has a link to has a link to http://blogs.msdn.com/b/fiddler/archive/2011/02/10/fiddler-is-better-with-internet-explorer-9.aspx but like most MS articles I come across these days it is no longer valid.

Considering this was a feature added in IE9, your testing indicates it doesn't work for WinHTTP, and the lack of docs for this in WinHTTP I will just assume that it isn't valid and remove the references to it in the docs for `win_http_proxy`.

I'll also update the documentation for `win_inet_proxy` to say you need IE9 for this to work.

johnboy2 commented:
Did some functional testing of these modules. Tried breaking them in a few different ways, but was mostly unsuccessful -- my tests all worked as expected _except_ for one. [See comment here](https://github.com/ansible/ansible/pull/54631/commits/bbf08126d2cf86ee7f262bce78bb264a9d91a89f#r271410503) for the one issue I found.

jborean93 committed:
Removed <-loopback> for win_http_proxy

jborean93 commented:
Thanks once again @johnboy2 for reviewing and testing this out. I've fixed the typos in the docs and removed mentions of `<-loopback>` for win_http_proxy.

jhawkesworth commented on lib/ansible/modules/windows/win_http_proxy.py:
```diff
@@ -0,0 +1,106 @@
+      or C(protocol=hostname:port).
+    - If the port is undefined, the default port for the protocol in use is
+      used.
+    - If setting a dict, the keys relate are the protocol and the value is the
```
Suggest rewording this sentence as 
"If setting a dict, the keys should be the protocol and the values should be the hostname and/or port for the protocol given in the key."

jhawkesworth commented on lib/ansible/modules/windows/win_inet_proxy.py:
```diff
@@ -0,0 +1,173 @@
+      or C(protocol=hostname:port).
+    - If the port is undefined, the default port for the protocol in use is
+      used.
+    - If setting a dict, the keys relate are the protocol and the value is the
```
Again, suggest rewording this as suggested for win_http_proxy.

jhawkesworth commented on lib/ansible/modules/windows/win_inet_proxy.py:
```diff
@@ -0,0 +1,173 @@
+    - Omit, set to null or an empty string to remove the proxy settings.
+notes:
+- This is not the same as the proxy settings set in WinHTTP through the
+  C(netsh) command. Use the M(win_http_proxy) module ot manage that instead.
```
typo: ot -> to

jhawkesworth commented:
@jborean93 - a couple of tiny suggestions for the docs.  I can't comment usefully on the native code but the rest of it looks good to me. I haven't tested though.

jborean93 committed:
doc changes from review

jborean93 commented:
Thanks @jhawkesworth for the review, have made the changes.

jborean93 commented:
Thanks to everyone for the review.
